test/test_query2.cpp

```diff
@@ -6056,44 +6056,93 @@ TEST(Query_FullTextMulti)
+    CHECK_EQUAL(do_fulltext_find("1990s c++"), Keys({4}));
+    CHECK_EQUAL(do_fulltext_find("object gemstone"), Keys({2, 4}));
 
     // over links
-    tv = origin->link(col_link).column<String>(col).fulltext("object gemstone").find_all();
-    CHECK_EQUAL(2, tv.size());
-
-    tv = origin->query("link.text TEXT 'object gemstone'").find_all();
-    CHECK_EQUAL(2, tv.size());
+    CHECK_EQUAL(do_query_find(origin, "link.text TEXT 'object gemstone'"), Keys({2, 4}));
+    auto tv = origin->link(col_link).column<String>(col).fulltext("object gemstone").find_all();
+    CHECK_EQUAL(get_keys(tv), Keys({2, 4}));
 
     // through LnkLst
     auto obj = tv.get_object(0);
     auto ll = obj.get_linklist(col_link);
     tv = table->where(ll).fulltext(col, "object gemstone").find_all();
-    CHECK_EQUAL(1, tv.size());
+    CHECK_EQUAL(get_keys(tv), Keys({2}));
 
     // Diacritics ignorant
-    tv = table->where().fulltext(col, "depot emanant").find_all();
-    CHECK_EQUAL(1, tv.size());
+    CHECK_EQUAL(do_fulltext_find("depot emanant archive"), Keys({5}));
 
     // search for combination that is not present
-    tv = table->where().fulltext(col, "object data").find_all();
-    CHECK_EQUAL(0, tv.size());
+    CHECK_EQUAL(do_fulltext_find("object data"), Keys());
+
+    // exclude words
+    CHECK_EQUAL(do_fulltext_find("-three one"), Keys({9}));
+    CHECK_EQUAL(do_fulltext_find("one -three"), Keys({9}));
+    CHECK_EQUAL(do_fulltext_find("object -databases"), Keys({1, 2, 4, 6}));
+    CHECK_EQUAL(do_fulltext_find("-databases object -duplicates"), Keys({1, 2, 4}));
+    CHECK_EQUAL(do_fulltext_find("object -objects"), Keys({2, 4, 6}));
+    CHECK_EQUAL(do_fulltext_find("-object objects"), Keys({3}));
+    CHECK_EQUAL(do_fulltext_find("databases -database"), Keys({}));
+    CHECK_EQUAL(do_fulltext_find("-database databases"), Keys({}));
+    CHECK_EQUAL(do_fulltext_find("database -databases"), Keys({1, 2, 4}));
+    CHECK_EQUAL(do_fulltext_find("-databases database"), Keys({1, 2, 4}));
+
+    // invalid exclude searches
+    // There should always be at least one mandatory term
+    CHECK_THROW_ANY(do_fulltext_find("-databases"));
+
+    // Token should only appear once
+    CHECK_THROW_ANY(do_fulltext_find("C# c++"));
+    CHECK_THROW_ANY(do_fulltext_find("-object object"));
+    CHECK_THROW_ANY(do_fulltext_find("object -object"));
+    CHECK_THROW_ANY(do_fulltext_find("objects -object object"));
+    CHECK_THROW_ANY(do_fulltext_find("object -object object"));
+    CHECK_THROW_ANY(do_fulltext_find("database -database"));
 
     // many terms
-    tv = table->where().fulltext(col, "object database management brown").find_all();
-    CHECK_EQUAL(1, tv.size());
-
-    tv = table->query("text TEXT 'object database management brown'").find_all();
-    CHECK_EQUAL(1, tv.size());
+    CHECK_EQUAL(do_fulltext_find("object database management brown"), Keys({1}));
+    CHECK_EQUAL(do_query_find(table, "text TEXT 'object database management brown'"), Keys({1}));
```

Comment:
```suggestion
    CHECK_EQUAL(do_query_find(table, "text TEXT 'object database management brown'"), Keys({1}));

    // FIXME tokenizer treats all these as separate tokens
    CHECK_EQUAL(do_fulltext_find("object-oriented"), Keys({0, 1}));
    CHECK_EQUAL(do_fulltext_find("object-oriented -databases"), Keys({1}));
    CHECK_EQUAL(do_fulltext_find("object-oriented -table-oriented"), Keys({1}));
```

These all currently fail, so the open question is if such single term like 'object-oriented' should be treated like two terms. Tokenizer probably doesn't fit fully into algo, and with exclusion it becomes even more confusing for what to expect from search.

Reply:
As stated above, you can't use the index to find words with embedded non alpha characters. If we should make this work, we would have to search the text as a post processing step. Definitely not within the scope of this PR